vulture_os/toolkit/api_parser/waf_barracuda/waf_barracuda.py

```diff
@@ -0,0 +1,127 @@
+
+        self.session = None
+
+    def __connect(self):
+
+        if not self.session:
+            self.session = requests.Session()
+            self.session.proxies = self.proxies
+            self.session.headers.update(self.HEADERS)
+            self.session.headers.update({"Authorization": f"Bearer {self.waf_barracuda_token}"})
+
+    def __execute_query(self, url, query={}, timeout=15):
+        response = self.session.get(url,
+            params=query,
+            headers=self.HEADERS,
+            timeout=timeout,
+            proxies=self.proxies,
+            verify=self.api_parser_custom_certificate if self.api_parser_custom_certificate else self.api_parser_verify_ssl
+        )
+        if response.status_code != 200:
+            raise WAFBarracudaAPIError(f"{[__parser__]}:get_logs: Status code = {response.status_code}, Error = {response.text}")
+        else:
+            return response.json()
+
+    def get_logs(self, since, to):
+        self.__connect()
+        result = []
+        query = {"from": since.timestamp(), "to": to.timestamp()}
+        logger.debug(f"{[__parser__]}:get_logs: params for request are {query}", extra={'frontend': str(self.frontend)})
+        url = f"https://{self.WAF_BARRACUDA_HOST}/v2/waasapi/applications/"
+        applications = self.__execute_query(url)
+        app_ids = [app.get("id") for app in applications.get("results")]
+        for app_id in app_ids:
+            url = f"https://{self.WAF_BARRACUDA_HOST}/v2/waasapi/applications/{app_id}/access/logs"
+            data = self.__execute_query(url, query=query)
+            result.extend(data["results"])
+            url = f"https://{self.WAF_BARRACUDA_HOST}/v2/waasapi/applications/{app_id}/waf/logs"
+            data = self.__execute_query(url, query=query)
+            result.extend(data["results"])
+        return result
+
+    def test(self):
+        try:
+            result = self.get_logs(timezone.now()-timedelta(days=1), timezone.now())
+
+            return {
+                "status": True,
+                "data": result
+            }
+        except Exception as e:
+            logger.exception(f"[{__parser__}]:test: {e}", extra={'frontend': str(self.frontend)})
+            return {
+                "status": False,
+                "error": str(e)
+            }
+
+    def format_log(self, log):
+        return json.dumps(log)
+
+    def execute(self):
+        try:
+            since = self.frontend.last_api_call or (timezone.now() - timedelta(days=1))
+            to = min(timezone.now(), since + timedelta(hours=24))
+
+            logger.info(f"[{__parser__}]:execute: getting logs from {since} to {to}", extra={'frontend': str(self.frontend)})
+            results = self.get_logs(since, to)
+            self.update_lock()
+
+            self.write_to_file([self.format_log(l) for l in results])
+            self.update_lock()
+            self.frontend.last_api_call = to
```

Comment:
what about duplicate logs? when a log timestamp corresponds to the 'to' value, won't it be fetched twice : once on the 'to' limit, and then on the 'since' limit?